### PR TITLE
Finish migration of Layer Presenter logic

### LINKED ISSUE #3674: Finish migration of Layer Presenter logic
- [x] `fix app/controllers/carto/api/layers_controller.rb` `load_parent` method loading a `@parent` that can be either a user or a map
- [x] replace `Layer::Presenter` for `LayerPresenter` at all carto/xxxx models/controllers
- [x] Inside `LayerPresenter`, change either all fields to use `@xxxx` or setup `attr_reader`, but no current mixture
- [x] After #3670 goes live, need to recheck all `LayerPresenter` calls
- [x] <del>replace `options[:viewer_user]` inside `layer_presenter.rb` for `current_viewer `</del>
- [x] Clean and refactor code inside `LayerPresenter`
- [x] Add specs for both presenters <del>and migrate `spec/models/layer/presenter_spec.rb`</del>

Nice to have:
- Check usages of 'user_name', now should rely on presenter options `:viewer_user` and `:user` -> [Moved to separate ticket](https://github.com/CartoDB/cartodb/issues/3745)

Branch: `3674-layer-presenter-migration`
PR: #3702

### PR DESCRIPTION
**Pull request by Kartones**
Fixes #3674

### CODE REVIEW

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3765/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3776/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3794/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3799/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3802/))

**Comment by Kartones** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3803/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3807/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3810/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3832/))

**Comment by Kartones** [comment]
retest this please

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3834/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3839/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3841/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3845/))

**Comment by Kartones** [comment]
@juanignaciosl CR please (I wrongly commented in the ticket and not here in the PR).

The `it_behaves_like` parameters might not be an awesome solution but does the job to reuse the same specs to run with different Presenters and Models so... at least works for now, can be improved in the future when we implement other CRUD operations in the refactored models.

I've backported some small bugfixes and tiny changes to add compatibility of `:viewer_user` etc. to old Presenter, but in general the sooner we switch to `Carto::LayerPresenter` the better.

File: app/controllers/carto/api/layer_presenter.rb [hunk 1]
```diff
@@ -4,102 +4,138 @@ module Carto
   module Api
     class LayerPresenter
 
-      PUBLIC_VALUES = [:kind, :infowindow, :tooltip, :id, :order, :parent_id]
+      PUBLIC_VALUES = %W{ options kind infowindow tooltip id order parent_id }
+
+      EMPTY_CSS = '#dummy{}'
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layer_presenter.rb]: Why is this needed? (just curious about it).

File: app/controllers/carto/api/layer_presenter.rb [hunk 2]
```diff
@@ -4,102 +4,138 @@ module Carto
...
-          order:      layer.order,
+          order:      @layer.order,
           options:    options_data_v1
         }
       end
 
       private
 
-      attr_reader :layer, :options, :configuration
+      def viewer_is_not_owner?
```
Comment (juanignaciosl) [inline app/controllers/carto/api/layer_presenter.rb]: This should probably be positive, `viewer_is_owner?`. And maybe even just `is_owner?(user)`.

File: app/controllers/carto/api/layer_presenter.rb (hunk 1 above)
Comment (Kartones) [inline app/controllers/carto/api/layer_presenter.rb]: The empty CSS is the template for when the asked layer type needs CSS styles and there's none present (model doesn't stores anything until the FE sends some CSS).

File: app/controllers/carto/api/layer_presenter.rb (hunk 2 above)
Comment (Kartones) [inline app/controllers/carto/api/layer_presenter.rb]: in negative the ifs were more clear to me, but as I've been intensively with this if you prefer positive and `is_owner` is ok too, I'll change it.

**Comment by juanignaciosl** [comment]
:+1:

File: app/controllers/carto/api/layer_presenter.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/layer_presenter.rb]: Ok, thanks. Maybe this comment might be useful inside the code.

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3846/))

**Comment by Cartofante** [comment]
Frontend tests were OK :+1: ([details](http://clinker.cartodb.net/jenkins/job/CartoDB-PR-testing/3848/))
